prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,1333 @@
+ *             1 otherwise
+ */
+static int util_find_overlapping_addr(void *x, void *y)
+{
+	struct util_mr_cache_key *to_find  = (struct util_mr_cache_key *)x;
+	struct util_mr_cache_key *to_compare = (struct util_mr_cache_key *)y;
+	uint64_t to_find_end = to_find->address + to_find->length;
+	uint64_t to_compare_end = to_compare->address + to_compare->length;
+
+	/* format: (x_addr,  x_len) - (y_addr,  y_len) truth_value
+	 *
+	 * case 1: (0x1000, 0x1000) - (0x1400, 0x0800) true
+	 * case 2: (0x1000, 0x1000) - (0x0C00, 0x0800) true
+	 * case 3: (0x1000, 0x1000) - (0x1C00, 0x0800) true
+	 * case 4: (0x1000, 0x1000) - (0x0C00, 0x2000) true
+	 * case 5: (0x1000, 0x1000) - (0x0400, 0x0400) false
+	 * case 6: (0x1000, 0x1000) - (0x2400, 0x0400) false
+	 */
+	if (!((to_find_end < to_compare->address) ||
+	      (to_compare_end < to_find->address)))
+		return 0;
+
+	/* left */
+	if (to_find->address < to_compare->address)
+		return -1;
+
+	return 1;
+}
+
+/**
+ * Key comparison function for memory registration caches
+ *
+ * @param[in] x key to be inserted or found
+ * @param[in] y key to be compared against
+ *
+ * @return    -1 if it should be positioned at the left, 0 if the same,
+ *             1 otherwise
+ */
+static inline int util_mr_cache_key_comp(void *x, void *y)
+{
+	struct util_mr_cache_key *to_insert  = (struct util_mr_cache_key *)x;
+	struct util_mr_cache_key *to_compare = (struct util_mr_cache_key *)y;
+
+	if (to_compare->address == to_insert->address)
+		return 0;
+
+	/* to the left */
+	if (to_insert->address < to_compare->address)
+		return -1;
+
+	/* to the right */
+	return 1;
+}
+
+/**
+ * Helper function for matching the exact key entry
+ *
+ * @param entry     memory registration cache key
+ * @param to_match  memory registration cache key
+ * @return 1 if match, otherwise 0
+ */
+static inline int util_match_exact_key(struct util_mr_cache_key *entry, 
+				    struct util_mr_cache_key *to_match)
+{
+	return (entry->address == to_match->address) &&
+	       (entry->length == to_match->length);
+}
+
+/**
+ * dlist search function for matching the exact memory registration key
+ *
+ * @param entry memory registration cache entry
+ * @param match memory registration cache key
+ * @return 1 if match, otherwise 0
+ */
+static inline int util_mr_exact_key(struct dlist_entry *entry,
+				    const void *match)
+{
+	struct util_mr_cache_entry *x =
+		container_of(entry, struct util_mr_cache_entry,
+			     siblings);
+	struct util_mr_cache_key *y = (struct util_mr_cache_key *)match;
+
+	return util_match_exact_key(&x->key, y);
+}
+
+/**
+ * Helper function to determine if one key subsumes another
+ *
+ * @param x  mr_cache_key
+ * @param y  mr_cache_key
+ * @return 1 if x subsumes y, 0 otherwise
+ */
+static inline int util_can_subsume(struct util_mr_cache_key *x,
+				   struct util_mr_cache_key *y)
+{
+	return (x->address <= y->address) &&
+			((x->address + x->length) >=
+					(y->address + y->length));
+}
```

Comment:
taking a quick look, seems it can't be replaced
the `util_mr_cache_check_overlapping_region` checks that there is an intersection between X region and Y region and returns 0 in a case of overlapping.
For example, the following two regions is overlapping regions, i.e. `util_mr_cache_check_overlapping_region` returns 0 for them (indicates that these are overlapping regions):
![image](https://user-images.githubusercontent.com/11650339/33312997-88697424-d43a-11e7-8047-ae171dab0e75.png)
But the `util_mr_cache_can_subsume` will return 0 for them that indicates inability to be subsumed.